charms/tensorboards-web-app/tests/unit/test_operator.py

```diff
@@ -0,0 +1,110 @@
+    harness.set_can_connect(APP_NAME, True)
+    return harness
+
+
+class TestCharm:
+    """Test class for TensorboardsWebApp."""
+
+    @patch("charm.KubernetesServicePatch", lambda x, y, service_name: None)
+    @patch("charm.TensorboardsWebApp.k8s_resource_handler")
+    def test_not_leader(self, k8s_resource_handler: MagicMock, harness: Harness):
+        """Test that charm waits if it's not the leader."""
+        harness.begin_with_initial_hooks()
+        assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+    @patch("charm.KubernetesServicePatch", lambda x, y, service_name: None)
+    @patch("charm.TensorboardsWebApp.k8s_resource_handler")
+    def test_no_relation(self, k8s_resource_handler: MagicMock, harness: Harness):
+        """Test that charm is blocked when there is no ingress relation."""
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
+        assert harness.charm.model.unit.status == BlockedStatus("No ingress relation available")
+
+    @patch("charm.KubernetesServicePatch", lambda x, y, service_name: None)
+    @patch("charm.TensorboardsWebApp.k8s_resource_handler")
+    def test_with_relation(self, k8s_resource_handler: MagicMock, harness: Harness):
+        """Test that charm is active when there is an ingress relation"""
+        harness.set_leader(True)
+        rel_id = self._setup_ingress_relation(harness)
+        harness.begin_with_initial_hooks()
+
+        assert harness.charm.model.unit.status == ActiveStatus("")
+        assert isinstance(harness.charm.model.unit.status, ActiveStatus)
+
+        relation_data = harness.get_relation_data(rel_id, harness.charm.app.name)
+        data = {
+            "service": "tensorboards-web-app",
+            "port": 5000,
+            "prefix": "/tensorboards",
+            "rewrite": "/",
+        }
+        assert data == yaml.safe_load(relation_data["data"])
+
+    @patch("charm.KubernetesServicePatch", lambda x, y, service_name: None)
+    @patch("charm.TensorboardsWebApp.k8s_resource_handler")
+    def test_pebble_layer(self, k8s_resource_handler: MagicMock, harness: Harness):
+        """Test the creation of Pebble layer and some of its fields."""
+        harness.set_leader(True)
+        harness.set_model_name("kubeflow")
+        self._setup_ingress_relation(harness)
+        harness.begin_with_initial_hooks()
+        assert harness.charm.container.get_service(APP_NAME).is_running()
+        pebble_plan = harness.get_container_pebble_plan(APP_NAME)
+        assert pebble_plan
+        assert pebble_plan.services
+        pebble_plan_info = pebble_plan.to_dict()
+        assert (
+            pebble_plan_info["services"][APP_NAME]["command"]
+            == "gunicorn -w 3 --bind 0.0.0.0:5000 --access-logfile - entrypoint:app"
+        )
+        test_env = pebble_plan_info["services"][APP_NAME]["environment"]
+        # there should be 7 environment variables
+        assert len(test_env) == 5
```

Comment:
7 or 5? Also, I wonder if checking the number of env variables here is scalable. I'm thinking what would happen if we decide to remove one of them.

Reply:
I always find these ones hard.  If you want to test something, I'd lean toward:
```python
for env_var in env_vars_i_expect:
  assert env_var in test_env
```

It confirms the things we expect right now are in there, and to @dnplas's point when someone in figure adds an environment variable it doesn't immediately break the test.  
... but someone could also argue it *should* break the test, so I dunno :D

Reply:
5 :) 

Breaking the environment variables test doesn't sound weird when we decide explicitly to modify the environments, since the test will also break if this happens unintentionally. A more meaningful change could be to add more assertions about each ENV variable we expect to be there. WDYT?

Reply:
to me, I say go with whatever you think is right but keep it low effort.  asserting that the right number is there, or maybe that the env vars you expect to exist are there is great.  I'd only check values if there's some logic that goes into them (eg: if `SOME_VAR=some-calculated-value` where `some-calculated-value` came from charm logic that you implemented)

Reply:
I agree with @ca-scribner 's:

```python
for env_var in env_vars_i_expect:
  assert env_var in test_env
```

This way we avoid the case where if you are testing just the number of vars (5), then you change one of them from say `var1` to `var2`, but the number of vars is still the same. This one looks low effort and even scalable.